digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/ProjectValidator.scala

```diff
@@ -423,6 +454,142 @@ object ProjectValidator {
+    * if they are not, issue a error
+    * @param project - the current project
+    * @param groupedProjectLinks - project links, grouped by road number and road part number
+    * @return Validation Errors
+    */
+  private def secondElyBorderCheck(project: RoadAddressProject, groupedProjectLinks: Map[(Long, Long), Seq[ProjectLink]]) = {
+    /**
+      * Method that will prepare the output of the validation error.
+      * @param validationError the validation error
+      * @param pl Sequence of the erroneous ProjectLinks
+      * @return An optional value with eventual Validation error details
+      */
+    def error(validationError: ValidationError)(pl: Seq[BaseRoadAddress]): Option[ValidationErrorDetails] = {
+      val grouppedByEly = pl.groupBy(_.ely)
+      val (gLinkIds, gPoints, gEly) = grouppedByEly.flatMap(g => {
+        val links = g._2
+        val zoomPoint = GeometryUtils.midPointGeometry(links.minBy(_.endAddrMValue).geometry)
+        links.map(l => (l.linkId, zoomPoint, l.ely))
+      }).unzip3
+
+      if (gLinkIds.nonEmpty)
+        Some(ValidationErrorDetails(project.id, validationError, gLinkIds.toSeq.distinct,
+          gPoints.map(p => ProjectCoordinates(p.x, p.y, 12)).toSeq.distinct, Some(RoadContinuesInAnotherElyMessage.format(gEly.toSet.mkString(", ")))))
+      else
+        Option.empty[ValidationErrorDetails]
+    }
+
+    /**
+      * Main validation we create a bounding box and search for adjacent road addresses to the edgeRoad.
+      * Then check if the ely code changed between them, if it did we output the edge road, if not, we output a empty optional
+      * @param edgeRoad - either the start or the end of a road number/road part number project link
+      * @return an optional symbolizing a found invalid edgeRoad, or nothing.
+      */
+    def checkEdgeValidation(edgeRoad: ProjectLink) = {
+      val p = endPoint(edgeRoad)
+      val lowerCorner = Point(p.x - 1.0, p.y - 1.0, p.z)
+      val higherCorner = Point(p.x + 1.0, p.y + 1.0, p.z)
+      val box = BoundingRectangle(lowerCorner, higherCorner)
+      val roadAddresses = RoadAddressDAO.fetchRoadAddressesByBoundingBox(box, false)
+        if(!roadAddresses.isEmpty) {
+          val filtered = roadAddresses.filterNot(ra => ra.roadNumber == edgeRoad.roadNumber && ra.roadPartNumber == edgeRoad.roadPartNumber &&
+            !GeometryUtils.areAdjacent(ra.geometry, edgeRoad.geometry))
+          if(filtered.find(_.ely != edgeRoad.ely).isDefined) {
+            Option(edgeRoad)
+          } else {
+            Option.empty[ProjectLink]
+          }
+        } else {
+          Option.empty[ProjectLink]
+        }
+    }
+
+    val validationProblems = groupedProjectLinks.flatMap(group => {
+      val projectLinks = group._2
+      val startRoad = projectLinks.head
+      val endRoad = projectLinks.last
+      val stv = if (startRoad.discontinuity.value != 3)checkEdgeValidation(startRoad) else Option.empty[ProjectLink]
```

Comment:
could be better to use "if (startRoad.discontinuity != Discontinuity.ChangingELYCode)"